torchmetrics/functional/text/ter.py

```diff
@@ -0,0 +1,617 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# referenced from
+# Library Name: torchtext
+# Authors: torchtext authors
+# Date: 2021-11-30
+# Link:
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+# Copyright 2020 Memsource
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from functools import lru_cache
+from typing import Dict, Iterator, List, Optional, Sequence, Tuple, Union
+
+from torch import Tensor, tensor
+
+from torchmetrics.functional.text.helper import (
+    _flip_trace,
+    _LevenshteinEditDistance,
+    _trace_to_alignment,
+    _validate_inputs,
+)
+
+# Tercom-inspired limits
+_MAX_SHIFT_SIZE = 10
+_MAX_SHIFT_DIST = 50
+
+# Sacrebleu-inspired limits
+_MAX_SHIFT_CANDIDATES = 1000
+
+
+class _TercomTokenizer:
+    """Re-implementation of Tercom Tokenizer in Python 3.
+
+    See src/ter/core/Normalizer.java in https://github.com/jhclark/tercom Note that Python doesn't support named Unicode
+    blocks so the mapping for relevant blocks was taken from here: https://unicode-table.com/en/blocks/
+
+    This implementation follows the implemenation from
+    https://github.com/mjpost/sacrebleu/blob/master/sacrebleu/tokenizers/tokenizer_ter.py.
+    """
+
+    _ASIAN_PUNCTUATION = r"([\u3001\u3002\u3008-\u3011\u3014-\u301f\uff61-\uff65\u30fb])"
+    _FULL_WIDTH_PUNCTUATION = r"([\uff0e\uff0c\uff1f\uff1a\uff1b\uff01\uff02\uff08\uff09])"
+
+    def __init__(
+        self,
+        normalize: bool = False,
+        no_punctuation: bool = False,
+        lowercase: bool = True,
+        asian_support: bool = False,
+    ) -> None:
+
+        """Initialize the tokenizer.
+
+        Args:
+            normalize:
+            An indication whether a general tokenization to be applied.
+        no_punctuation:
+            An indication whteher a punctuation to be removed from the sentences.
+        lowercase:
+            An indication whether to enable case-insesitivity.
+        asian_support:
+            An indication whether asian characters to be processed.
+        """
+        self.normalize = normalize
+        self.no_punctuation = no_punctuation
+        self.lowercase = lowercase
+        self.asian_support = asian_support
```

Comment:
Please implement input validation here. Then args are correctly validated for both functional and modular